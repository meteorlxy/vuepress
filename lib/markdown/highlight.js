const chalk = require('chalk')
const prism = require('prismjs')
const loadLanguages = require('prismjs/components/index')
const escapeHtml = require('escape-html')

const vPreRE = /^(.*):v$/

// required to make embedded highlighting work...
loadLanguages(['markup', 'css', 'javascript'])

function wrap (code, lang, useVPre = true) {
  if (lang === 'text') {
    code = escapeHtml(code)
  }
  const vPreAttr = useVPre ? 'v-pre' : ''
  return `<pre ${vPreAttr} class="language-${lang}"><code>${code}</code></pre>`
}

module.exports = (str, lang) => {
  const vPreMatch = lang.match(vPreRE)
  const useVPre = vPreMatch === null
  if (!useVPre) {
    lang = vPreMatch[1]
  }
  if (!lang) {
    return wrap(str, 'text', useVPre)
  }
  lang = lang.toLowerCase()
  const rawLang = lang
  if (lang === 'vue' || lang === 'html') {
    lang = 'markup'
  }
  if (lang === 'md') {
    lang = 'markdown'
  }
  if (!prism.languages[lang]) {
    try {
      loadLanguages([lang])
    } catch (e) {
      console.log(chalk.yellow(`[vuepress] Syntax highlight for language "${lang}" is not supported.`))
    }
  }
  if (prism.languages[lang]) {
    const code = prism.highlight(str, prism.languages[lang], lang)
    return wrap(code, rawLang, useVPre)
  }
  return wrap(str, 'text', useVPre)
}
