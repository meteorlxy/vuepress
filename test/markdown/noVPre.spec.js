import { Md, getFragment } from './util'
import highlight from '@/markdown/highlight.js'

const md = Md().set({ highlight })

describe('no-v-pre', () => {
  test('should not have v-pre with {no-v-pre}', async () => {
    const input = await getFragment('code-no-v-pre')
    const output = md.render(input)
    expect(output).toMatchSnapshot()
  })
})
