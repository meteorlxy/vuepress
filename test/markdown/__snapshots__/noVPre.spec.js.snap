// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`no-v-pre should not have v-pre with {no-v-pre} 1`] = `
<pre class="language-vue"><code>{{ 1 + 1 }}
</code></pre>
`;
